eng/common/scripts/get-codeowners.ps1

```diff
@@ -1,49 +1,49 @@
 param (
-  $TargetDirectory, # should be in relative form from root of repo. EG: sdk/servicebus
-  $RootDirectory, # ideally $(Build.SourcesDirectory)
-  $VsoVariable = "" # target devops output variable
+  [string]$CodeOwnerPathExpression, # Code path to code owners. e.g sdk/core/azure-amqp
+  [string]$ToolVersion = "", # Placeholder. Will update in next PR
+  [string]$ToolPath = "$env:AGENT_TOOLSDIRECTORY", # The place to check the tool existence. Put $(Agent.ToolsDirectory) as default
+  [string]$DevOpsFeed = "https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json", # DevOp tool feeds.
+  [string]$CodeOwnerFileLocation = "$env:SYSTEM_DEFAULTWORKINGDIRECTORY/.github/CODEOWNERS", # The place we fetch CODEOWNERS file
+  [string]$VsoVariable = "" # Option of write code owners into devop variable
 )
-$target = $TargetDirectory.ToLower().Trim("/")
-$codeOwnersLocation = Join-Path $RootDirectory -ChildPath ".github/CODEOWNERS"
-$ownedFolders = @{}
 
-if (!(Test-Path $codeOwnersLocation)) {
-  Write-Host "Unable to find CODEOWNERS file in target directory $RootDirectory"
-  exit 1
-}
-
-$codeOwnersContent = Get-Content $codeOwnersLocation
 
-foreach ($contentLine in $codeOwnersContent) {
-  if (-not $contentLine.StartsWith("#") -and $contentLine){
-    $splitLine = $contentLine -split "\s+"
-    
-    # CODEOWNERS file can also have labels present after the owner aliases
-    # gh aliases start with @ in codeowners. don't pass on to API calls
-    $ownedFolders[$splitLine[0].ToLower().Trim("/")] = ($splitLine[1..$($splitLine.Length)] `
-      | ? { $_.StartsWith("@") } `
-      | % { return $_.substring(1) }) -join ","
+# Check if the retrieve-codeowners tool exsit or not.
+function InstallRetrieveCodeOwnersTool() {
+  if (Get-Command "$ToolPath/retrieve-code-owners") {
+    return
   }
+  dotnet tool install --tool-path $toolPath --add-source "$DevOpsFeed" --version "$ToolVersion" "Azure.Sdk.Tools.RetrieveCodeOwners"
 }
 
-$results = $ownedFolders[$target]
+# Get the json property from tool command.
+function GetCodeOwnersEntryFromCommand() {
+  return & "$ToolPath/retrieve-code-owners" --target-directory "$CodeOwnerPathExpression" --code-owner-file-path "$CodeOwnerFileLocation" 
+}
 
-if ($results) {
-  Write-Host "Found a folder $results to match $target"
+function GetCodeOwners([string]$codeOwnersString) 
+{ 
+  $codeOwnersJson = $codeOwnersString | ConvertFrom-Json
+  if (!$codeOwnersJson) {
+    Write-Host "No code owners returned from the path: $CodeOwnerPathExpression"
+    return ""
+  }
```

Comment:
Nit: it's not a huge deal here, but as a larger principal, this log statement uses knowledge of how the underlying `GetCodeOwnersEntryFromCommand` function is implemented (using $CodeOwnerPathExpression), a function which isn't even called by this function. It would be better to log this message from within the `GetCodeOwnersEntryFromCommand` function since it's specific to its implementation.

I would prefer to delete these lines and `throw` from the `GetCodeOwnersEntryFromCommand` function instead, which makes the code simpler.

Reply:
We fairly get into this condition.
If tool command return error or json with context. 
```
if (codeOwnerEntry == null)
{
    Console.Error.WriteLine(String.Format("We cannot find any closest code owners from the target path {0}", targetDirectory));
    return 1;
}
else
{
    var codeOwnerJson = JsonSerializer.Serialize<CodeOwnerEntry>(codeOwnerEntry);
    Console.WriteLine(codeOwnerJson);
    return 0;
}
```
The error handling here is for this line 
```
$codeOwnersJson = $codeOwnersString | ConvertFrom-Json
```

However, instead of checking null or empty, it is better to do error handling.

Reply:
My thinking is we could set `$ErrorActionPreference` to `Stop` for the script context to handle these cases since `ConvertFrom-Json` will throw. If the caller has this set in their shell, they'll skip your log statement anyway.